### DISCUSSION TITLE
How can I change the number of channels of a pretrained model?

### DISCUSSION THREAD

**Discussion by aykutcayir34** [PyTorchLightning/lightning-flash#740] [Q&A]
I want to use a single channel resnet-18 model in flashlightning image classifier. Is there a convenient way to change the number of channels of a pre-trained model in flashlightning?

**Comment by ethanwharris** [comment]
Hi @aykutcayir34 you should be able to do this:
```
model = ImageClassifier(
    backbone="resnet18",
    backbone_kwargs={"in_chans": 1},
    ...
)
```

Hope that helps :smiley:

**Reply by aykutcayir34** [reply]
Nope, I have tried the backbone keyword argument like you suggested above. Unfortunately, the solution does not work :(

**Reply by ethanwharris** [reply]
Could you share the error you get? We have an example of it working for an efficient net here: https://www.kaggle.com/ethanwharris/seti-meets-lightning-flash?rvi=1&scriptVersionId=71922551&cellId=7

It's possible that it didn't work for resnets in previous versions, could you also share which version of flash you have?

**Reply by aykutcayir34** [reply]
I got this:

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_9060/3607665525.py in <module>
----> 1 model = ImageClassifier(backbone="resnet18", num_classes=1, backbone_kwargs={"in_chans": 1})

~\anaconda3\envs\flashlightningimgaud\lib\site-packages\flash\image\classification\model.py in __init__(self, num_classes, backbone, backbone_kwargs, head, pretrained, loss_fn, optimizer, optimizer_kwargs, scheduler, scheduler_kwargs, metrics, learning_rate, multi_label, serializer)
    111             self.backbone, num_features = backbone
    112         else:
--> 113             self.backbone, num_features = self.backbones.get(backbone)(pretrained=pretrained, **backbone_kwargs)
    114 
    115         head = head(num_features, num_classes) if isinstance(head, FunctionType) else head

~\anaconda3\envs\flashlightningimgaud\lib\site-packages\flash\core\utilities\url_error.py in wrapper(pretrained, *args, **kwargs)
     22     def wrapper(*args, pretrained=False, **kwargs):
     23         try:
---> 24             return fn(*args, pretrained=pretrained, **kwargs)
     25         except urllib.error.URLError:
     26             # Hack for icevision/efficientdet to work without internet access

~\anaconda3\envs\flashlightningimgaud\lib\site-packages\flash\image\classification\backbones\resnet.py in _resnet(model_name, block, layers, num_features, pretrained, weights_paths, **kwargs)
    352 
    353     if model_weights is not None:
--> 354         backbone.load_state_dict(model_weights)
    355 
    356     return backbone, num_features

~\anaconda3\envs\flashlightningimgaud\lib\site-packages\torch\nn\modules\module.py in load_state_dict(self, state_dict, strict)
   1221 
   1222         if len(error_msgs) > 0:
-> 1223             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
   1224                                self.__class__.__name__, "\n\t".join(error_msgs)))
   1225         return _IncompatibleKeys(missing_keys, unexpected_keys)

RuntimeError: Error(s) in loading state_dict for ResNet:
	size mismatch for conv1.weight: copying a param with shape torch.Size([64, 3, 7, 7]) from checkpoint, the shape in current model is torch.Size([64, 1, 7, 7]).

**Reply by aykutcayir34** [reply]
flash version is '0.5.0rc0'

**Reply by aykutcayir34** [reply]
Your solution worked for efficientnet, but I could not run the solution for resnet-18 :/

**Reply by ethanwharris** [reply]
Hi @aykutcayir34 thanks for the extra information! This is definitely a bug on our side. The efficientnet comes from timm underneath which has a way to handle pretrained weights with different input channels. Our resnet code is a little different so we need to add support for this there. I've opened #741 so that you can track the progress of this issue. In the meantime, the efficient net and any resenet variants like resnext or other models should all work. The resnet18 should work ok if you set pretrained to false.

Hope that helps :smiley:

**Reply by aykutcayir34** [reply]
Thank you so much. You helped me a lot at speed light ;)

**Reply by ethanwharris** [reply]
@aykutcayir34 you're welcome :smiley: This PR should fix your issue once it's merged: https://github.com/PyTorchLightning/lightning-flash/pull/743
